### PR TITLE
Update to wayland-rs 0.21

### PR DESCRIPTION
**Pull request by vberger**
This upgrades smithay to wayland-rs 0.21, fixing a lot of the issues of the wayland backend, and removing a lot of unsafe code.

This changed slightly the API of backends: some arguments could no longer easily be handed back on failure. However none of them seemed to be used nor necessarily useful, so I just removed them.

@Drakulix can you have a look at my changes on the backend side? Maybe just look at the first commit, the second one is just a cargo fmt.

### CODE REVIEW

File: src/backend/session/direct.rs [hunk 1]
```diff
@@ -397,18 +398,47 @@ impl Implementation<(), SignalEvent> for DirectSessionNotifier {
...
+/// Bound logind session that is driven by the `wayland_server::EventLoop`.
+///
+/// See `direct_session_bind` for details.
+pub struct BoundDirectSession {
+    source: Source<Signals>,
+    notifier: Rc<RefCell<DirectSessionNotifier>>
+}
+
+impl BoundDirectSession {
+    /// Unbind the logind session from the `EventLoop`
```
Comment (Drakulix) [inline src/backend/session/direct.rs]: Copy-Paste error in the comment

**Comment by vberger** [comment]
> While you are at it, would you mind telling me why `LoopHandle` has associated data now and how that should be used?

Sure, it's something that was suggested to me to allow more generic uses cases of `calloop`. The iead is that when calling `event_loop.dispatch(...)`, you provide it with a `&mut Data` argument, which is forwarded to all callbacks as a second argument. To allow easy data sharing between callbacks if needed.

As of this PR, smithay does not use it, but it'd be possible to integrate some constraints like `where Data: SomeTrait` in a module that'd need it, I guess. Though I think in general it's more targeted to end applications. I'd expect intermediate libs like smithay or wayland-rs to either ignore it (like this PR does as best as it can), or forwarding it to the user callbacks.

File: src/backend/libinput.rs [hunk 2]
```diff
@@ -600,26 +600,19 @@ impl<S: Session> libinput::LibinputInterface for LibinputSessionInterface<S> {
...
-                if let Err(error) = self.dispatch_new_events() {
-                    warn!(self.logger, "Libinput errored: {}", error);
-                }
-            }
-            FdEvent::Error { error, .. } => {
-                warn!(self.logger, "Libinput fd errored: {}", error);
+pub fn libinput_bind<Data: 'static>(
+    mut backend: LibinputInputBackend,
+    handle: LoopHandle<Data>,
+) -> ::std::result::Result<Source<Generic<EventedRawFd>>, IoError> {
```
[the remark this answers is not in the document]
Comment (Drakulix) [reply src/backend/libinput.rs]: It is ignored, but in theory you could run the backend manually. This is fine for libinput, as you can just create another backend, but I don't really like the API inconsistency here.

**Comment by Drakulix** [comment]
Thanks for clearing that up. I am not sure, if we should fix the inconsistency in the return value API. Maybe we can build a little helper to make it easier to return the old values?

I would generally prefer to give the user control over what ffi functions are run and dropping the `Libinput` context struct, does call into libinput and makes the creation of a new context necessary, another call that may fail. However I do not have very strong feelings about this approach.

**Comment by vberger** [comment]
Alright, I'll try to see if I can refactor a little the modules to make returning the backend on error possible in a not too ugly way.

**Comment by vberger** [comment]
Alright, merging, travis is happy enough. :)

**Pull request merged**